src/realm/object-store/sync/app.cpp

```diff
@@ -1222,123 +1180,128 @@ void App::do_request(Request&& request, UniqueFunction<void(const Response& resp
+                                                              "Redirect response missing location header",
+                                                              response.http_status_code));
     }
 
     // Request the location info at the new location - once this is complete, the original
     // request will be sent to the new server
     update_location_and_resend(std::move(request), std::move(completion), std::move(redir_location));
 }
 
-void App::do_authenticated_request(Request&& request, const std::shared_ptr<User>& sync_user,
+void App::do_authenticated_request(HttpMethod method, std::string&& route, std::string&& body,
+                                   const std::shared_ptr<User>& user, RequestTokenType token_type,
                                    util::UniqueFunction<void(const Response&)>&& completion)
 {
-    request.headers = get_request_headers(sync_user, request.uses_refresh_token ? RequestTokenType::RefreshToken
-                                                                                : RequestTokenType::AccessToken);
-
-    auto completion_2 = [completion = std::move(completion), request, sync_user,
-                         self = shared_from_this()](const Response& response) mutable {
+    auto request = make_request(method, std::move(route), user, token_type, std::move(body));
+    do_request(std::move(request), [token_type, user, completion = std::move(completion), self = shared_from_this()](
+                                       std::unique_ptr<Request>&& request, const Response& response) mutable {
         if (auto error = AppUtils::check_for_errors(response)) {
-            self->handle_auth_failure(std::move(*error), std::move(response), std::move(request), sync_user,
+            self->handle_auth_failure(std::move(*error), std::move(request), response, user, token_type,
                                       std::move(completion));
         }
         else {
             completion(response);
         }
-    };
-    do_request(std::move(request), std::move(completion_2));
+    });
 }
 
-void App::handle_auth_failure(const AppError& error, const Response& response, Request&& request,
-                              const std::shared_ptr<User>& sync_user,
+void App::handle_auth_failure(const AppError& error, std::unique_ptr<Request>&& request, const Response& response,
+                              const std::shared_ptr<User>& user, RequestTokenType token_type,
                               util::UniqueFunction<void(const Response&)>&& completion)
 {
     // Only handle auth failures
-    if (*error.additional_status_code == 401) {
-        if (request.uses_refresh_token) {
-            if (sync_user && sync_user->is_logged_in()) {
-                sync_user->log_out();
-            }
-            completion(response);
-            return;
-        }
+    if (*error.additional_status_code != 401) {
+        completion(response);
+        return;
     }
```

Comment:
The logic here is unchanged and it's just rearranged slightly because I found it hard to understand when exactly we were retrying. In both versions of the code we retry only if the error code is 401 and the request did not use the refresh token. We did not previously retry on non-401 errors.

Reply:
Oh - you're right, I got a bit confused with the diff here and thought it would always fall through to the `refresh_access_token()` underneath if an error occurred.